modules-js/react-fleet/src/form-elements/inputs/MemorableDateInput.tsx

```diff
@@ -0,0 +1,458 @@
+import React from 'react';
+
+import { css } from 'emotion';
+
+interface Props {
+  initialDate?: string;
+  earliestDate?: string;
+  latestDate?: string;
+  onlyAllowPast?: boolean;
+  onlyAllowFuture?: boolean;
+  legend: React.ReactChild;
+  handleDate: (newDate: Date | null) => void;
+}
+
+interface State {
+  hasFocus: boolean;
+  error: string;
+  date: Date | null;
+  fields: {
+    month: number;
+    day: number;
+    year: number;
+  };
+}
+
+/**
+ * Component to collect a memorable date (such as a birth date) from a user.
+ *
+ * An element must be passed in as the legend, commonly a header <h1-h6>.
+ *
+ * Default behavior is to accept any date. A text string representing a date
+ * can be passed into earliestDate and/or latestDate to limit acceptable input.
+ * Format: MM-DD-YYYY, MM/DD/YYYY, etc. (https://tools.ietf.org/html/rfc2822#section-3.3)
+ *
+ * An initial date can be passed in via the initialDate property.
+ *
+ * onlyAllowPast: If true, will only accept yesterday or earlier for a date.
+ * onlyAllowFuture: If true, will only accept tomorrow or later for a date.
+ *
+ * Whenever the date is updated, this.props.handleDate is called.
+ *
+ * *Please note that all dates are relative to the user’s computer.
+ *
+ * Inspired by https://design-system.service.gov.uk/components/date-input/
+ */
+export default class MemorableDateInput extends React.Component<Props, State> {
+  readonly earliest: Date | null;
+  readonly latest: Date | null;
+  readonly initial: Date | null;
+
+  private timeout: any;
+
+  constructor(props: Props) {
+    super(props);
+
+    this.initial = props.initialDate ? new Date(props.initialDate) : null;
+
+    checkIfPropsValid(props);
+
+    this.earliest = setDateLimit(
+      props.earliestDate || '',
+      !!props.onlyAllowFuture
+    );
+    this.latest = setDateLimit(props.latestDate || '', !!props.onlyAllowPast);
+
+    this.state = {
+      hasFocus: false,
+      error: '',
+      date: this.initial,
+      fields: this.initial
+        ? {
+            month: this.initial.getMonth(),
+            day: this.initial.getDate(),
+            year: this.initial.getFullYear(),
+          }
+        : ({
+            month: '',
+            day: '',
+            year: '',
+          } as any),
+    };
+  }
+
+  // Determine if date range limits have been declared, and whether a given
+  // date is within that range. If not, this.state.error is set to reflect this.
+  private checkIfDateInRange(date: Date): boolean {
+    const errorString = isDateValid(
+      date,
+      this.earliest,
+      this.latest,
+      this.props.onlyAllowPast,
+      this.props.onlyAllowFuture
+    );
+
+    this.setState({ error: errorString });
+
+    return errorString.length === 0;
+  }
+
+  // If date is within range, update value. Otherwise, clear the value.
+  private updateDate(): void {
+    let date: Date | null = new Date(
```

Comment:
Given that it’s unclear what locale is being used for this parsing, it might be better to make the `Date` using the `(year, month [, date [, hours [, minutes [, seconds [, ms ] ] ] ] ] )` arguments. (You have numbers anyway.)